### ISSUE TITLE
Grimblast on hyprland with a monitor scaling above 1 results in blurry screenshots

### ISSUE THREAD

**Issue by IceDBorn** [hyprwm/contrib#69]
# Screenshot with scaling set to 1
![image](https://github.com/hyprwm/contrib/assets/51162078/78c4f80b-2778-4bef-97c9-851769e57781)

# Screenshot with scaling set to 1.15
![image](https://github.com/hyprwm/contrib/assets/51162078/fd169dff-1839-4aaf-9726-9fdca94d1171)

**Comment by fufexan** [comment]
I think that's the application not properly supporting the fractional scaling protocol, and being blurred itself.

**Comment by IceDBorn** [comment]
Nope, the app was crystal clear upon screenshotting. I even used a magnifier app to replicate the screenshot and it was like the first screenshot where the scaling is set to 1.

**Comment by IceDBorn** [comment]
I use the option, but I do not get any blurriness with it when the scaling is set to 1.

**Comment by IceDBorn** [comment]
Nope

**Comment by anotherwanwanyi** [comment]
I'm experiencing a similar issue.
I use a laptop connected to a monitor.
![图片](https://github.com/hyprwm/contrib/assets/72936236/34317f99-9c25-431d-9084-e814daa429b2)
As picture shows, my laptop screen scaling is set to 2, while the monitor scaling is set to 1.5.

It seems like the blur comes from the "--freeze" option.

**eDP-1(Scaling 2) Screenshot with `-f(--freeze)`**:
![图片](https://github.com/hyprwm/contrib/assets/72936236/090ea9d2-5976-4c14-b483-40ed773e7a62)

**eDP-1(Scaling 2) Screenshot without `-f(--freeze)`**:
![图片](https://github.com/hyprwm/contrib/assets/72936236/7e8026c5-5121-4500-a70a-8ded98c32d0a)

**HDMI-A-1(Scaling 1.5) Screenshot with `-f(--freeze)`**:
![图片](https://github.com/hyprwm/contrib/assets/72936236/257716a7-9f65-4e72-87bb-c6cc3caebd4a)

**HDMI-A-1(Scaling 1.5) Screenshot without `-f(--freeze)`**:
![图片](https://github.com/hyprwm/contrib/assets/72936236/eb5cbbf4-8151-4c7b-81e7-474b2441aeaa)

The third and fourth images are both blurry, but the third one is blurrier. I suspect the blur in the fourth image is due to the 1.5x scaling, while the blur in the third one is also influenced by the `--freeze` option.

Another more intuitive way to verify this is to simply set the scaling to a non-integer value (e.g., 1.5) and bind the keys as follows: 
```
bind = , Print,                     exec, grimblast -n -f copy area
```
Just press "Print" again and again, and the blur will overlap continuously.
However, using `grimblast -n copy area` won't cause that blur.

**Comment by fufexan** [comment]
If it's the freeze option's fault, then hyprpicker is at fault. @vaxerski

**Comment by vaxerski** [comment]
it's not hyprpicker, it's hyprland. On fs screenshots of an area are blurry. That's why freeze makes it _more_ blurred, because blurring happens twice. (hyprpicker is 1, grim is 2)

**Comment by purrpurrn** [comment]
Why do hyprpicker and grim blur the image?

**Comment by decoded-satapathy** [comment]
I have expireinced the same issue, with scaling set to integer values (1, 2, 3....) I get a sharper screenshot, while with non integer scaling (1.25, 1.5 ...) I get visibly blurry screenshot


**Screenshot with scaling set to 1.25**
![image](https://github.com/hyprwm/contrib/assets/115103314/49accc9e-e2b9-48b7-9acd-ca47a79b63c2)

**Screenshot with scaling set to 1**
![image](https://github.com/hyprwm/contrib/assets/115103314/d1b086dc-1680-4680-a471-a18ff468b2c5)

I haven't used the --freeze option.

**Comment by ZerdoX-x** [comment]
We already confirmed the issue:

> it's not hyprpicker, it's hyprland. On fs screenshots of an area are blurry. That's why freeze makes it more blurred, because blurring happens twice. (hyprpicker is 1, grim is 2)

please do not bump with "me too"

**Comment by KAGEYAM4** [comment]
Any workaround for this other than not using fractional scaling?

**Comment by Sewer56** [comment]
There's a patch https://lists.sr.ht/~emersion/grim-dev/patches/56912 that works on my end after backporting to current `grim`. 

Unfortunately, despite being a simple patch; it's been sitting there for over a year.

**Comment by thequantumcog** [comment]
I refactored the above patch to make it work on latest master
https://raw.githubusercontent.com/thequantumcog/grim-patch/refs/heads/main/hyprland.patch

**Comment by uxobsessed** [comment]
Has this been merged?

> I refactored the above patch to make it work on latest master https://raw.githubusercontent.com/thequantumcog/grim-patch/refs/heads/main/hyprland.patch
